src/components/interface/cbuf_mgr/cos_alloc.c

```diff
@@ -0,0 +1,401 @@
+static void do_cbuf_free(cbuf_t cbid)
+{
+    cbuf_free(cbid);
+}
+
+static void _alloc_libc_free(void *ptr) 
+{
+	if (ptr) {
+		__alloc_t *pointer = BLOCK_START(ptr);
+		size_t size = pointer->size;
+
+		if (size) {
+			cbuf_free(pointer->cbuf_id);
+		}
+	}
+}
+
+void free(void *ptr) __attribute__((weak,alias("_alloc_libc_free")));
+
+#ifdef WANT_MALLOC_ZERO
+static __alloc_t zeromem[2];
+#endif
+
+static void* _alloc_libc_malloc(size_t size) {
+  __alloc_t* ptr;
+  size_t need;
+#ifdef WANT_MALLOC_ZERO
+  if (!size) return BLOCK_RET(zeromem);
+#else
+  if (!size) goto err_out;
+#endif
+  size+=sizeof(__alloc_t);
+  if (unlikely(size<sizeof(__alloc_t))) goto err_out;
+
+  if (size<=__MAX_SMALL_SIZE) {
+    need=GET_SIZE(size);
+    ptr=__small_malloc(need);
+  } 
+  else {
+    need=PAGE_ALIGN(size);
+    cbuf_t cbid;
+    ptr = need ? do_cbuf_alloc(need, &cbid) : MAP_FAILED;
+    ptr->cbuf_id = cbid;
+  }
+  if (ptr==MAP_FAILED) goto err_out;
+  ptr->size=need;
+  return BLOCK_RET(ptr);
+err_out:
+  return 0;
+}
+
+void* malloc(size_t size) __attribute__((alias("_alloc_libc_malloc")));
+
+void *__libc_calloc(size_t nmemb, size_t _size)
+{
+	size_t tot = nmemb*_size;
+	char *ret = malloc(tot);
+	
+	memset(ret, 0, tot);
+	return ret;
+}
+
+void* calloc(size_t nmemb, size_t _size) __attribute__((weak,alias("__libc_calloc")));
+
+/* gabep1 additions for allocations of pages. */
+
+void *alloc_page(void)
+{
+	struct free_page *fp;
+	void *a;
+
+	fp = page_list.next;
+	if (NULL == fp) {
+		cbuf_t cbid;
+		a = do_cbuf_alloc(PAGE_SIZE, &cbid);
+	} 
+	else {
+		page_list.next = fp->next;
+		fp->next = NULL;
+		a = (void*)fp;
+	}
+	
+	return a;
+}
+
+void free_page(void *ptr)
+{
+	struct free_page *fp;
+	
+	fp = (struct free_page *)ptr;
+	fp->next = page_list.next;
+	page_list.next = fp;
+
+	return;
+}
+
+#ifdef NIL
+
+void* __libc_calloc(size_t nmemb, size_t _size);
+void* __libc_calloc(size_t nmemb, size_t _size) {
+  register size_t size=_size*nmemb;
+  if (nmemb && size/nmemb!=_size) {
+    (*__errno_location())=ENOMEM;
+    return 0;
+  }
+  return malloc(size);
+}
+void* calloc(size_t nmemb, size_t _size) __attribute__((weak,alias("__libc_calloc")));
+
+void* __libc_realloc(void* ptr, size_t _size);
+void* __libc_realloc(void* ptr, size_t _size) {
+  register size_t size=_size;
+  if (ptr) {
+    if (size) {
+      __alloc_t* tmp=BLOCK_START(ptr);
+      size+=sizeof(__alloc_t);
+      if (size<sizeof(__alloc_t)) goto retzero;
+      size=(size<=__MAX_SMALL_SIZE)?GET_SIZE(size):PAGE_ALIGN(size);
+      if (tmp->size!=size) {
+	if ((tmp->size<=__MAX_SMALL_SIZE)) {
+	  void *new=_alloc_libc_malloc(_size);
+	  if (new) {
+	    register __alloc_t* foo=BLOCK_START(new);
+	    size=foo->size;
+	    if (size>tmp->size) size=tmp->size;
+	    if (size) memcpy(new,ptr,size-sizeof(__alloc_t));
+	    _alloc_libc_free(ptr);
+	  }
+	  ptr=new;
+	}
+	else {
+	  register __alloc_t* foo;
+	  size=PAGE_ALIGN(size);
+	  foo=mremap(tmp,tmp->size,size,MREMAP_MAYMOVE);
```

Comment:
This will cause a problem, right?  Or did I miss this somehow being implemented with cbufs.